editor/tree_search.cpp

```diff
@@ -0,0 +1,492 @@
+		result.upper = MAX(result.upper, static_cast<int>(word_position + word.length()));
+	}
+
+	return result;
+}
+
+void TreeSearch::_select_item(TreeItem *p_item) {
+	if (!p_item)
+		return;
+	ERR_FAIL_COND(!tree_reference || p_item->get_tree() != tree_reference);
+
+	// first unfold ancestors
+	TreeItem *ancestor = p_item->get_parent();
+	while (ancestor) {
+		ancestor->set_collapsed(false);
+		ancestor = ancestor->get_parent();
+	}
+	//then scroll to [item]
+	tree_reference->scroll_to_item(p_item);
+
+	// ...and select it
+	tree_reference->deselect_all();
+	tree_reference->set_selected(p_item, 0);
+}
+
+void TreeSearch::_select_first_match() {
+	if (matching_entries.size() == 0) {
+		return;
+	}
+	for (int i = 0; i < ordered_tree_items.size(); i++) {
+		TreeItem *item = ordered_tree_items[i];
+		if (!_vector_has_bsearch(matching_entries, item)) {
+			continue;
+		}
+		String debug_string = "[";
+		_select_item(item);
+		return;
+	}
+}
+
+void TreeSearch::_select_next_match() {
+	if (matching_entries.size() == 0) {
+		return;
+	}
+	TreeItem *selected = tree_reference->get_selected(); // we care about a single item here.
+	if (!selected) {
+		_select_first_match();
+		return;
+	}
+
+	int selected_idx = -1;
+	for (int i = 0; i < ordered_tree_items.size(); i++) {
+		if (ordered_tree_items[i] == selected) {
+			selected_idx = i;
+			break;
+		}
+	}
+
+	// find the best fitting entry.
+	for (int i = 0; i < ordered_tree_items.size(); i++) {
+		TreeItem *item = ordered_tree_items[i];
+		if (!_vector_has_bsearch(matching_entries, item) || selected_idx >= i) {
+			continue;
+		}
+
+		_select_item(item);
+		return;
+	}
+	_select_first_match(); // wrap around.
+}
+
+template <typename T>
+inline bool TreeSearch::_vector_has_bsearch(Vector<T *> p_vec, T *element) {
+	int idx = p_vec.bsearch(element, true);
+	bool in_array = idx >= 0 && idx < p_vec.size();
+
+	return in_array && p_vec[idx] == element;
+}
+
+void TreeSearch::on_item_edited(TreeItem *item) {
+	if (item->get_cell_mode(0) != TreeItem::CELL_MODE_CUSTOM) {
+		return;
+	}
+
+	if (!callable_cache.has(item) || item->get_custom_draw_callback(0) == callable_cache.get(item)) {
+		return;
+	}
+
+	item->set_custom_draw_callback(0, callable_cache.get(item));
+}
+
+// Call this as a post-processing step for the already constructed tree.
+void TreeSearch::update_search(Tree *p_tree) {
+	ERR_FAIL_COND(!search_panel || !p_tree);
+
+	// ignore if panel not visible or no search string is given.
+	if (!search_panel->is_visible() || search_panel->get_text().length() == 0) {
+		return;
+	}
+
+	tree_reference = p_tree;
+
+	String search_mask = search_panel->get_text();
+	TreeSearchMode search_mode = search_panel->get_search_mode();
+
+	_update_ordered_tree_items(p_tree->get_root());
+	_update_matching_entries(search_mask);
+	_update_number_matches();
+
+	_highlight_tree(search_mask);
+	if (!search_panel->is_connected("text_submitted", callable_mp(this, &TreeSearch::_select_next_match))) {
```

Comment:
Should be in `NOTIFICATION_READY`

Reply:
Since this is `RefCounted`, I now put it into the constructor. is_connected needs no checking this way. Great catch!

Reply:
Done.